### PR TITLE
add additional special functions (spherical harmonics, Legendre, Gamma functions)

### LINKED ISSUE #5684: [feature request] implement cupyx.scipy.special.sph_harm
Hi all,

it would be great to have a cupyx.scipy.special.sph_harm function as in https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.sph_harm.html

Cheers,
Peter

### PR DESCRIPTION
**Pull request by grlee77**
closes #5684

Marked as draft because:

~- a couple of test cases related to `gammaincinv` and `gammainccinv` are failing (unexpected NaN values)~ Fixed by 
17c75f4  
~- we may need to expand the tests a bit more.~ -> new `sph_harm` and `log1p` test cases have been added
- I wanted to think about how to potentially refactor. For example, notice that the ufunc code template for `sph_harm` includes the template strings for `lpmv` which in turn includes other functions, etc.. Thus, there is duplicate C code across the CUDA kernels, but I'm not sure of the easiest way to avoid that. 

I reorganized the `__init__.py` a bit to group related functions together (as in the SciPy docs). The new functions added in this PR are:
`sph_harm`, `log1p`, `gamminc`, `gammincinv`, `gammincc`, `gammainccinv`, `gammasgn`, `lpmv` and `poch`.

The additional functions are here because they were needed within the implementation of `sph_harm`.

### CODE REVIEW

**Comment by grlee77** [comment]
The test failures have now been resolved and additional tests for `sph_harm` and `log1p` have been added.

**Comment by emcastillo** [comment]
seems the tests are failing with cuda 10.2 in windows :(

File: cupyx/scipy/special/_lpmv.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    }
+    return ga;
+}
+
+
+/*
+ *     Purpose: Compute the associated Legendre function
+ *              Pmv(x) with an integer order and an arbitrary
+ *              nonnegative degree v
+ *     Input :  x   --- Argument of Pm(x)  ( -1 ≤ x ≤ 1 )
```
Comment (emcastillo) [inline cupyx/scipy/special/_lpmv.py:174]: I think windows might not be liking ≤

File: cupyx/scipy/special/_lpmv.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    }
+    if (x < 0.) {
+        ps = ps - PI*cos(PI*x)/sin(PI*x) - 1.0/x;
+    }
+    return ps;
+}
+
+
+/*
+ *       Purpose: Compute gamma function Г(x)
```
Comment (emcastillo) [inline cupyx/scipy/special/_lpmv.py:101]: The Г char here is breaking windows CI :)

**Review by emcastillo — COMMENTED** [review]
LGTM but we need to change some of the unicode characters for the CIs to work.
Thanks!

File: cupyx/scipy/special/_gammainc.py [hunk 3]
```diff
@@ -0,0 +1,1202 @@
...
+# See the following on error states:
+#    docs.scipy.org/doc/scipy/reference/generated/scipy.special.errstate.html
+#    scipy/special/sf_error.h
+
+# TODO: which value for MACHEP, MAXLOG from const.c to use?
+
+_misc_preamble = """
+
+// include for CUDART_NAN, CUDART_INF
+#include <math_constants.h>
```
Comment (leofang) [inline cupyx/scipy/special/_gammainc.py:43]: See https://github.com/cupy/cupy/pull/5687#issuecomment-907355030.
```suggestion
#include <cupy/math_constants.h>
```

File: cupyx/scipy/special/_poch.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+poch_definition = (
+    lgam_definition
+    + gammasgn_definition
+    + """
+/*
+ * Pochhammer symbol (a)_m = gamma(a + m) / gamma(a)
+ */
+
+// include for CUDART_NAN, CUDART_INF
+#include <math_constants.h>
```
Comment (leofang) [inline cupyx/scipy/special/_poch.py:27]: Ditto
```suggestion
#include <cupy/math_constants.h>
```

File: cupyx/scipy/special/_sph_harm.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+sph_harmonic_definition = (
+    poch_definition
+    + lpmv_definition
+    + """
+
+#include <cupy/complex.cuh>
+
+// include for CUDART_NAN, CUDART_INF
+#include <math_constants.h>
```
Comment (leofang) [inline cupyx/scipy/special/_sph_harm.py:21]: Ditto.
```suggestion
#include <cupy/math_constants.h>
```

**Review by leofang — CHANGES_REQUESTED** [review]

**Review by emcastillo — APPROVED** [review]

File: cupyx/scipy/special/_gammainc.py [hunk 6]
```diff
@@ -0,0 +1,1202 @@
...
+    /* 1.0000000000000000000000E0, */
+    1.5062909083469192043167E1,
+    8.3047565967967209469434E1,
+    2.2176239823732856465394E2,
+    3.0909872225312059774938E2,
+    2.1642788614495947685003E2,
+    6.0118660497603843919306E1,
+};
+
+__device__ double log1p(double x)
```
Comment (asi1024) [inline cupyx/scipy/special/_gammainc.py:134]: Can this function be replaced by CUDA [log1p](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__DOUBLE.html#group__CUDA__MATH__DOUBLE_1g3c680d660d75780ef53075a439211626) or [log1pf](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1g9d53128ab5f7d6ebc4798f243481a6d7)?

Comment (grlee77) [reply cupyx/scipy/special/_gammainc.py:134]: Yes, I tested that the CUDA Math `log1p` still passes the tests. I can make a small PR to drop the redundant code

**Pull request merged**